spark/src/main/scala/org/apache/spark/sql/delta/files/TransactionalWrite.scala

```diff
@@ -449,4 +469,27 @@ trait TransactionalWrite extends DeltaLogging { self: OptimisticTransactionImpl
 
     resultFiles.toSeq ++ committer.changeFiles
   }
+
+  /**
+   * Optimized writes can be enabled/disabled through the following order:
+   *  - Through DataFrameWriter options
+   *  - Through the table parameter
+   *  - Through SQL configuration
+   */
+  private def shouldOptimizeWrite(
+      writeOptions: Option[DeltaOptions], sessionConf: SQLConf): Boolean = {
+    writeOptions.flatMap(_.optimizeWrite)
+      .getOrElse(TransactionalWrite.shouldOptimizeWrite(metadata, sessionConf))
+  }
+
+}
+
+object TransactionalWrite {
+  def shouldOptimizeWrite(metadata: Metadata, sessionConf: SQLConf): Boolean = {
+    // We want table properties to take precedence over the session/default conf.
+    DeltaConfigs.OPTIMIZE_WRITE
+      .fromMetaData(metadata)
+      .orElse(sessionConf.getConf(DeltaSQLConf.DELTA_OPTIMIZE_WRITE_ENABLED))
```

Comment:
When session config is set, shouldn't it have precedence over table properties?

Reply:
makes sense